### PR TITLE
Feature/#10 train svm with labeled acorn data

### PR DESCRIPTION
**Pull request by Qbicz**
- gathered computed Hu moments in a single file
- manually labelled the pictures of acorns
- trained SVM

Effect: a hyperplane separating 7D space of Hu moments
TODO next: test the SVM using testing dataset
TODO: tune Canny threshold to detect more objects from the prepared images

### CODE REVIEW

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp [hunk 1]
```diff
@@ -8,57 +8,70 @@
 
 #include "print_vector.h"
 
-#define DATA_DIMENSIONS 2 // for Hu moments, it will be 7
+#define DATA_DIMENSIONS 7 // there are 7 Hu moments
```
Comment (palucki) [inline PC/svm/svm_train_and_save/svm_train_and_save.cpp]: Why not to be modern and use `constexpr int DATA_DIMENSIONS = 7;` ? :D

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp [hunk 2]
```diff
@@ -8,57 +8,70 @@
...
+    std::cout << "Training data set: " << training_data_json.dump(4) << std::endl;
     std::cout << "Data set size: " << training_data_json.size() << std::endl;
 
     // Set up training data
     std::vector<float> labels;
     std::vector<float> training_data;
+    unsigned int labeled_dataset_size = 0;
 
+    // For each acorn read its learning label and Hu moments
     for (auto element : training_data_json)
```
Comment (palucki) [inline PC/svm/svm_train_and_save/svm_train_and_save.cpp]: we can use `const auto& element : ...` as long as we do not modify it ;)

File: modules/print_vector.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 void print_vector(std::vector<float> &vect);
 std::string vector_to_string(std::vector<float> &vect);
+std::vector<float> string_to_vector(std::string str);
```
Comment (palucki) [inline modules/print_vector.h]: please pass const std::string& str

File: PC/hu_moments_gather/hu_moments_gather.cpp [hunk 4]
```diff
@@ -76,6 +76,7 @@ int main(int argc, char** argv)
 
         std::cout << "Hu moments string: " << hu_moments_string << std::endl;
 
+        hu_json["image"] = image_file;
         hu_json["hu_moments"] = hu_moments_string;
```
Comment (palucki) [inline PC/hu_moments_gather/hu_moments_gather.cpp:80]: how about adding hierarchical structure?
We can index hu moments with simple:
```
    "hu_moments": {
        "hu1":"...",
        "hu2":"...",
        "hu3":"..."
    }
```
will it make futher computations easier?

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp [hunk 5]
```diff
@@ -8,57 +8,70 @@
...
-    training_element.push_back(element["y"]);
-    return training_element;
-}
-
 int main()
 {
-    // Read input training data to JSON object
-    std::ifstream input_training_data("../data/dataset_training.json");
+    // Prepare file with input training data
+    std::ifstream input_training_data("../data/dataset_training_labeled_canny150.json");
```
Comment (palucki) [inline PC/svm/svm_train_and_save/svm_train_and_save.cpp]: Do we want to recompile with every change in input data?
Maybe we can pass it as CLI arg?

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp [hunk 6]
```diff
@@ -8,57 +8,70 @@
...
+    // Read training data to JSON object
     json training_data_json;
     input_training_data >> training_data_json;
- 
-    std::cout << "Training data set: " << training_data_json << std::endl;
+
+    std::cout << "Training data set: " << training_data_json.dump(4) << std::endl;
```
Comment (palucki) [inline PC/svm/svm_train_and_save/svm_train_and_save.cpp]: What is that magic number? ;)

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp (hunk 1 above)
Comment (Qbicz) [reply PC/svm/svm_train_and_save/svm_train_and_save.cpp]: Is there any difference between const and constexpr for a variable? Compiler knows const int is a compile time constant.

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp (hunk 6 above)
Comment (Qbicz) [reply PC/svm/svm_train_and_save/svm_train_and_save.cpp]: Fixed

File: PC/svm/svm_train_and_save/svm_train_and_save.cpp (hunk 2 above)
Comment (Qbicz) [reply PC/svm/svm_train_and_save/svm_train_and_save.cpp]: Good point, this is less copying. Fixed

File: modules/print_vector.h (hunk 3 above)
Comment (Qbicz) [reply modules/print_vector.h]: Done

**Comment by palucki** [comment]
According to this:
https://stackoverflow.com/questions/13346879/const-vs-constexpr-on-variables
 in our situation there is no big diff. Can be left as is now.

**Comment by palucki** [comment]
But, wait a minute. Without constexpr we can not perform `static_assert`
which I see you prefer ;)

**Pull request merged**